### PR TITLE
Add a TCP example to the repository

### PR DESCRIPTION
**Pull request by Bhashinee**
## Purpose
This is to add a sample to the TCP repository. This will demonstrate a simple HTTP echo request/response cycle written using the TCP package.
Fixes: https://github.com/ballerina-platform/ballerina-standard-library/issues/1781

## Examples

After running the client the console output will look like the below,

```
POST /test HTTP/1.1
Host:foo.example
Content-Type:text/plain
custom-header:header1
Content-Length: 12

request body

HTTP/1.1 200 Ok
Content-Type:text/plain
custom-header:header1
Content-Length: 12

request body
```

## Checklist
- [x] Linked to an issue
- [ ] Updated the changelog
- [x] Added tests

### CODE REVIEW

File: examples/http-client-server/client/client.bal [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+};
+
+type Response record {
+    map<string> headers;
+    string body;
+};
+
+Response response = {
+    headers: {},
+    body: ""
```
Comment (shafreenAnfar) [inline examples/http-client-server/client/client.bal:40]: Why have we assigned default values here ? instead can't we use nillable ?

File: examples/http-client-server/client/client.bal [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+};
+public function main() returns error? {
+    tcp:Client socketClient = check new ("localhost", 3000);
+
+    string requestString = createRequest(request);
+    io:println(requestString);
+    check socketClient->writeBytes(requestString.toBytes());
+
+    readonly & byte[] receivedData = check socketClient->readBytes();
```
Comment (shafreenAnfar) [inline examples/http-client-server/client/client.bal:49]: Ideally what we need is, first parse the headers and then look for the content length and exit the loop once the content length is reached.

Comment (Bhashinee) [reply examples/http-client-server/client/client.bal:47]: Changed the logic accordingly with https://github.com/ballerina-platform/module-ballerina-tcp/pull/379/commits/db402d3353480e33374c960af17bacd40bc2f25f

File: examples/http-client-server/client/client.bal [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+
+type Request record {
+    map<string> headers;
+    string body;
+    string HttpVersion = HTTP_1_1;
+    string Method = POST;
+    string path;
+};
+
+enum Method {
```
Comment (shafreenAnfar) [inline examples/http-client-server/client/client.bal:30]: Let's keep the `enums` together.

File: examples/http-client-server/client/client.bal [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+import ballerina/io;
+import ballerina/lang.'int as ints;
+import ballerina/regex;
+import ballerina/tcp;
+
+type Request record {
```
Comment (shafreenAnfar) [inline examples/http-client-server/client/client.bal:22]: Let's keep it next to the `Response` record.

File: examples/http-client-server/server/server.bal [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+
+type Request record {
+    map<string> headers;
+    string? body;
+    string HttpVersion = HTTP_1_1;
+    string Method = POST;
+    string path;
+};
+
+enum Method {
```
Comment (shafreenAnfar) [inline examples/http-client-server/server/server.bal:34]: ditto

File: examples/http-client-server/server/server.bal [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+type Response record {
+    map<string> headers;
+    string? body = ();
+    string status;
+    string HttpVersion;
+};
+
+service on new tcp:Listener(3000) {
+    remote function onConnect(tcp:Caller caller) returns tcp:ConnectionService {
+        return new EchoService();
```
Comment (shafreenAnfar) [inline examples/http-client-server/server/server.bal:52]: EchoHttpService

File: examples/http-client-server/server/server.bal [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+                match self.status {
+                    WAITING => {
+                        error? result = self.parseInitialChunk(request); 
+                    }
+                    RECEIVING_BODY => {
+                        error? result = self.parseBody(data);
+                    }
+                }
+                if self.status == RECEIVED_BODY {
+                    self.createRequestRecord();
```
Comment (shafreenAnfar) [inline examples/http-client-server/server/server.bal:89]: Let's parse the values explicitly, it increases the readability. 

```ballerina
Request req = self.createRequestRecord(self.headersMap, self.payload, self.httpStatus, self.httpVersion);
byte[] response = self.serializeResponse(self.headersMap, self.payload, self.httpStatus, self.httpVersion, req);
```

File: examples/http-client-server/server/server.bal [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+import ballerina/regex;
+import ballerina/io;
+
+enum stateMachine {
+    WAITING, RECEIVING_BODY, RECEIVED_BODY
+}
+
+type Request record {
+    map<string> headers;
+    string? body;
```
Comment (shafreenAnfar) [inline examples/http-client-server/server/server.bal:28]: Let's do the same as `Response` string? body = ();

File: examples/http-client-server/client/client.bal (hunk 3 above)
Comment (Bhashinee) [reply examples/http-client-server/client/client.bal:22]: Fixed all the comments https://github.com/ballerina-platform/module-ballerina-tcp/pull/379/commits/df79651d3030fce4bb9c405de1257c506c417440

File: examples/http-client-server/server/server.bal [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+    private int payloadLength = 0;
+    private string httpVersion = "HTTP/1.1";
+    private string httpStatus = "200 Ok";
+
+    remote isolated function onBytes(tcp:Caller caller, readonly & byte[] data) returns tcp:Error? {
+        string|error request = string:fromBytes(data);
+        lock {
+            self.httpStatus = request is error ? "404 Bad Request" : "200 Ok";
+        }
+        if (request is error) {
```
Comment (shafreenAnfar) [inline examples/http-client-server/server/server.bal:70]: We can write it as follows right,

```ballerina
if request is error
```

Comment (Bhashinee) [reply examples/http-client-server/server/server.bal]: Fixed with https://github.com/ballerina-platform/module-ballerina-tcp/pull/379/commits/1599b6dfa51ac438fb4098f1237142abffb5d78e

**Review by shafreenAnfar — APPROVED** [review]

**Pull request merged**